ledgers/ccf/transaction_processor/pdo_tp.cpp

```diff
@@ -54,26 +55,70 @@ namespace ccfapp
         };
 
         //======================================================================================================
-        // register enclave expected measurements (member method)
-        auto set_contract_enclave_attestatation_verification_policy = [this](auto& ctx, const nlohmann::json& params) {
+        // register contract enclave attestation check flag (member method)
+        auto set_contract_enclave_attestation_check_flag = [this](auto& ctx, const nlohmann::json& params) {
 
-            const auto in = params.get<RegisterContractEnclaveAttestationVerificationPolicy::In>();
+            const auto in = params.get<RegisterContractEnclaveAttestionCheckFlag::In>();
 
-            // get the current measurements view
-            auto attestation_policy_view = ctx.tx.rw(attestation_policy_table);
+            // get the current view of contract_enclave_check_attestation_flag
+            auto check_attestation_flag_view = ctx.tx.rw(contract_enclave_check_attestation_flag);
+
+            // Current PDO policy permits the flag to be set only once. Check if already set.
+
+            // Below we check the ccf node has a local copy of the attestation flag. If yes, an error
+            // is returned. (Note that global commit of the flag might be pending, and this is OK).
+            auto check_attestation_flag_check = check_attestation_flag_view->get(PDO_ENCLAVE_CHECK_ATTESTATION_FLAG);
```

Comment:
It's not clear why it "is OK" to access a local copy of the flag, instead of the globally committed one.
"Pending" does not mean that the transaction "will be" committed.
In fact, it could turn out to be invalid -- https://microsoft.github.io/CCF/main/use_apps/verify_tx.html

Should this be switched to `get_globally_committed`?

Reply:
Local read is what we need IMHO. All that we do here is prevent re-registration of the flag as long as local commit has been made. If you look at the enclave registration rpc, there we ensure that until the flag has been globally committed, enclave registration will not be permitted. 

The pattern is used at other places within the CCF TP as well. (local reads are used to check that a variable does not exist, while global reads are used when the variable should exist). 

Yes, a local->global commit can fail, in which case we will continue to see failures in the rpc when we try to use the variable. Fixes are tricky IMO, and depends on the reasons for the underlying failure. For example, if the failure is because the RAFT primary failed before global commit, then it is as good as saying that "the set flag rpc" was never issued, and in this case, we can simply re-issue the "set flag rpc" after we find that one of the CCF nodes failed.

Like you say, the failure of global commit can happen due to some other CCF internal reason as well. In these cases, we must have mechanism to resubmit the transaction (if we need solutions agnostic to the failure reason). Making the whole CCF TP robust to possible CCF commit failures should be flagged as a separate Issue, beyond the scope of this PR.

Checking for global values directly is not an option, since this allows multple end-points to "sucessfully" submit the set flag rpcs.

Reply:
It's a longer discussion, and I have started issue #473 to gather/discuss this information.